Lib/importlib/_bootstrap_external.py

```diff
@@ -209,7 +209,9 @@ def _write_atomic(path, data, mode=0o666):
         # We first write data to a temporary file, and then use os.replace() to
         # perform an atomic rename.
         with _io.FileIO(fd, 'wb') as file:
-            file.write(data)
```

Comment:
The write can be partial for a number of reasons, and while some of those are retried (PEP-0475), things like "this is windows and the write size is capped" aren't retried. Could this retry, which hopefully should also mean the OSError contains underlying errno / error info?

FileIO.write calls os.write calls _Py_write calls _Py_write_impl:

https://github.com/python/cpython/blob/6293d00e7201f3f28b1f4512e57dc4f03855cabd/Python/fileutils.c#L1918-L2016

Reply:
I don't know whether we should retry here. in a sense that's a separate question to the bug that I'm trying to fix. we should definitely not silently write a truncated pyc file, as we are right now.

Reply:
My thought is “why didn’t the write error itself” and doing a second one / retry might fix that

Reply:
because `write` doesn't error, it just writes fewer bytes. from the man page:

>  The number of bytes written may be less than count if, for example, there is insufficient space on the underlying physical medium, or the RLIMIT_FSIZE resource limit is encountered (see setrlimit(2)), or the call
>  was interrupted by a signal handler after having written less than count bytes.  (See also pipe(7).)

https://man7.org/linux/man-pages/man2/write.2.html#DESCRIPTION

Reply:
Also from that though, a second call won't return 0, either it will write write more and rerturn the count, or it will error (which should make the OSError): 

> In the event of a partial write, the caller can make another write() call to transfer the remaining bytes.  The subsequent call will either transfer further bytes or may result in an error (e.g., if the disk is now full).

Reply:
Considering this issue hasn't been reported until now I don't think it's worth complicating the code to try and be clever to work around some odd OS restriction.